File: src/backend/src/NextStep.Domain/AthleteProfile/Entities/Athlete.cs
using NextStep.Domain.AthleteProfile.ValueObjects;
using NextStep.Domain.Common;

namespace NextStep.Domain.AthleteProfile.Entities;

public class Athlete : Entity<Guid>
{
    public PersonalInfo PersonalInfo { get; private set; } = null!;
    public PhysiologicalData PhysiologicalData { get; private set; } = null!;
    public TrainingAccess TrainingAccess { get; private set; } = null!;
    public TrainingAvailability TrainingAvailability { get; private set; } = null!;

    private readonly List<Goal> _goals = new();
    public IReadOnlyList<Goal> Goals => _goals.AsReadOnly();

    public IReadOnlyList<HeartRateZone> HeartRateZones => CalculateHeartRateZones();
    public IReadOnlyList<PaceZone> PaceZones => CalculatePaceZones();

    private Athlete() : base() { }

    private Athlete(Guid id, PersonalInfo personalInfo) : base(id)
    {
        PersonalInfo = personalInfo;
        PhysiologicalData = PhysiologicalData.Empty();
        TrainingAccess = TrainingAccess.Default();
        TrainingAvailability = TrainingAvailability.Default();
    }

    public static Athlete Create(string name, DateOnly birthDate)
    {
        var personalInfo = PersonalInfo.Create(name, birthDate);
        return new Athlete(Guid.NewGuid(), personalInfo);
    }

    public static Athlete CreateWithId(Guid id, string name, DateOnly birthDate)
    {
        var personalInfo = PersonalInfo.Create(name, birthDate);
        return new Athlete(id, personalInfo);
    }

    public void UpdatePersonalInfo(string name, DateOnly birthDate)
    {
        PersonalInfo = PersonalInfo.Create(name, birthDate);
    }

    public void UpdatePhysiologicalData(int? maxHeartRate, int? lactateThresholdHeartRate, TimeSpan? lactateThresholdPace)
    {
        PhysiologicalData = PhysiologicalData.Create(maxHeartRate, lactateThresholdHeartRate, lactateThresholdPace);
    }

    public void UpdateTrainingAccess(bool hasTrackAccess)
    {
        TrainingAccess = TrainingAccess.Create(hasTrackAccess);
    }

    public void UpdateTrainingAvailability(
        WorkoutType monday,
        WorkoutType tuesday,
        WorkoutType wednesday,
        WorkoutType thursday,
        WorkoutType friday,
        WorkoutType saturday,
        WorkoutType sunday)
    {
        TrainingAvailability = TrainingAvailability.Create(
            monday, tuesday, wednesday, thursday, friday, saturday, sunday);
    }

    public Goal AddGoal(DateOnly raceDate, TimeSpan targetTime, GoalDistance distance)
    {
        var isPrimary = _goals.Count == 0;
        var goal = Goal.Create(raceDate, targetTime, distance, isPrimary);
        _goals.Add(goal);
        return goal;
    }

    public void UpdateGoal(Guid goalId, DateOnly raceDate, TimeSpan targetTime, GoalDistance distance)
    {
        var goal = _goals.FirstOrDefault(g => g.Id == goalId)
            ?? throw new InvalidOperationException($"Goal with id {goalId} not found");

        goal.Update(raceDate, targetTime, distance);
    }

    public void DeleteGoal(Guid goalId)
    {
        var goal = _goals.FirstOrDefault(g => g.Id == goalId)
            ?? throw new InvalidOperationException($"Goal with id {goalId} not found");

        if (goal.IsPrimary && _goals.Count > 1)
        {
            throw new InvalidOperationException("Cannot delete primary goal when other goals exist. Set another goal as primary first.");
        }

        _goals.Remove(goal);
    }

    public void SetPrimaryGoal(Guid goalId)
    {
        var goal = _goals.FirstOrDefault(g => g.Id == goalId)
            ?? throw new InvalidOperationException($"Goal with id {goalId} not found");

        foreach (var g in _goals)
        {
            g.SetPrimary(g.Id == goalId);
        }
    }

    private IReadOnlyList<HeartRateZone> CalculateHeartRateZones()
    {
        if (!PhysiologicalData.LactateThresholdHeartRate.HasValue)
            return Array.Empty<HeartRateZone>();

        return HeartRateZone.CalculateFromLactateThreshold(PhysiologicalData.LactateThresholdHeartRate.Value);
    }

    private IReadOnlyList<PaceZone> CalculatePaceZones()
    {
        if (!PhysiologicalData.LactateThresholdPace.HasValue)
            return Array.Empty<PaceZone>();

        return PaceZone.CalculateFromLactateThreshold(PhysiologicalData.LactateThresholdPace.Value);
    }
}

File: src/frontend/src/types/athlete.ts
export interface Athlete {
  id: string;
  personalInfo: PersonalInfo;
  physiologicalData: PhysiologicalData;
  trainingAccess: TrainingAccess;
  trainingAvailability: TrainingAvailability;
  heartRateZones: HeartRateZone[];
  paceZones: PaceZone[];
  goals: Goal[];
}

export interface PersonalInfo {
  name: string;
  birthDate: string;
  age: number;
}

export interface PhysiologicalData {
  maxHeartRate: number | null;
  lactateThresholdHeartRate: number | null;
  lactateThresholdPace: string | null;
}

export interface TrainingAccess {
  hasTrackAccess: boolean;
}

export interface TrainingAvailability {
  monday: WorkoutType;
  tuesday: WorkoutType;
  wednesday: WorkoutType;
  thursday: WorkoutType;
  friday: WorkoutType;
  saturday: WorkoutType;
  sunday: WorkoutType;
}

export type WorkoutType =
  | 'Rest'
  | 'CrossHIIT'
  | 'Recovery'
  | 'EasyRun'
  | 'Speed'
  | 'TempoRun'
  | 'LongRun';

export const WORKOUT_TYPES: { value: WorkoutType; label: string; isQuality: boolean }[] = [
  { value: 'Rest', label: 'Rest', isQuality: false },
  { value: 'CrossHIIT', label: 'Cross / HIIT', isQuality: true },
  { value: 'Recovery', label: 'Recovery', isQuality: false },
  { value: 'EasyRun', label: 'Easy Run', isQuality: false },
  { value: 'Speed', label: 'Speed', isQuality: true },
  { value: 'TempoRun', label: 'Tempo Run', isQuality: true },
  { value: 'LongRun', label: 'Long Run', isQuality: true },
];

export interface HeartRateZone {
  zoneNumber: number;
  name: string;
  minBpm: number;
  maxBpm: number;
}

export interface PaceZone {
  zoneNumber: number;
  name: string;
  minPace: string;
  maxPace: string;
}

export interface CreateAthleteRequest {
  name: string;
  birthDate: string;
}

export interface UpdatePersonalInfoRequest {
  name: string;
  birthDate: string;
}

export interface UpdatePhysiologicalDataRequest {
  maxHeartRate: number | null;
  lactateThresholdHeartRate: number | null;
  lactateThresholdPace: string | null;
}

export interface UpdateTrainingAccessRequest {
  hasTrackAccess: boolean;
}

export interface UpdateTrainingAvailabilityRequest {
  monday: number;
  tuesday: number;
  wednesday: number;
  thursday: number;
  friday: number;
  saturday: number;
  sunday: number;
}

export interface Goal {
  id: string;
  raceDate: string;
  targetTime: string;
  distanceType: DistanceType;
  customDistanceKm: number | null;
  distanceDisplay: string;
  isPrimary: boolean;
}

export type DistanceType =
  | 'Distance1600m'
  | 'Distance5K'
  | 'Distance10K'
  | 'Distance16K'
  | 'HalfMarathon'
  | 'Marathon'
  | 'Custom';

export const DISTANCE_TYPES: { value: DistanceType; label: string; km: number | null }[] = [
  { value: 'Distance1600m', label: '1600m', km: 1.6 },
  { value: 'Distance5K', label: '5K', km: 5 },
  { value: 'Distance10K', label: '10K', km: 10 },
  { value: 'Distance16K', label: '16K', km: 16 },
  { value: 'HalfMarathon', label: 'Half Marathon', km: 21.0975 },
  { value: 'Marathon', label: 'Marathon', km: 42.195 },
  { value: 'Custom', label: 'Custom', km: null },
];

export interface AddGoalRequest {
  raceDate: string;
  targetTime: string;
  distanceType: number;
  customDistanceKm: number | null;
}

export interface UpdateGoalRequest {
  raceDate: string;
  targetTime: string;
  distanceType: number;
  customDistanceKm: number | null;
}
